File: HW/HW11/Main.py
class Solution:
    def floodFill(self, image, sr, sc, newColor):
        color = image[sr][sc]
        if color == newColor: 
            return image
        def dfs(r, c):
            if image[r][c] == color:
                image[r][c] = newColor
                if r >= 1: dfs(r-1, c)
                if r+1 < len(image): dfs(r+1, c)
                if c >= 1: dfs(r, c-1)
                if c+1 < len(image[0]): dfs(r, c+1)

        dfs(sr, sc)
        return image

File: HW/HW11/Main.java
public class Main {
    private int[][] DIRECTIONS = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        if (image[sr][sc] != newColor)  {
            dfs(image, sr, sc, image[sr][sc], newColor);
        }
        return image;
    }

    private void dfs(int[][] image, int row, int column, int color, int newColor) {
        if (row < 0 || row >= image.length || column < 0 || column >= image[0].length || image[row][column] != color) {
            return;
        }

        image[row][column] = newColor;

        for (int[] direction : DIRECTIONS) {
            dfs(image, row + direction[0], column + direction[1], color, newColor);
        }
    }
}